lib/shoulda/matchers/active_record/have_attached_matcher.rb

```diff
@@ -0,0 +1,133 @@
+module Shoulda
+  module Matchers
+    module ActiveRecord
+      def have_one_attached(name)
+        HaveAttachedMatcher.new(:one, name)
+      end
+
+      def have_many_attached(name)
+        HaveAttachedMatcher.new(:many, name)
+      end
+
+      # @private
+      class HaveAttachedMatcher
+        attr_reader :name
+
+        def initialize(macro, name)
+          @macro = macro
+          @name = name
+        end
+
+         def description
+          "has_#{macro}_attached #{name}"
+        end
+
+        def failure_message
+          "Expected #{expectation} (#{@failure})"
+        end
+
+        def failure_message_when_negated
+          "Did not expect #{expectation}"
+        end
+
+        def expectation
+          "#{model_class.name} to have a has_#{macro} attached called #{name}"
+        end
+
+        def matches?(subject)
+          @subject = subject
+          reader_attribute_exists? &&
+            writer_attribute_exists? &&
+            attachments_association_exists? &&
+            blobs_association_exists? &&
+            eager_loading_scope_exists?
+        end
+
+        private
+
+        attr_reader :subject, :macro
+
+        def attachments_association_exists?
```

Comment:
Can you reorganize these methods so that they appear in the same order that they're called in `matches?` That makes it easier to follow.

Reply:
Done. I've put the sub-methods (so the private methods needed for a check for example `attachments_association_name`) directly below the associated methods. Let me know if you want to move them down so the methods of `matches?` follow each other directly.